### ISSUE TITLE
[RFC] Enable Int8-Mixed-BF16 PT2E PTQ Quantization with Inductor

### ISSUE THREAD

**Issue by leslie-fang-intel** [pytorch/pytorch#111640]
### 🚀 The feature, motivation and pitch

[Intel-Extension-for-PyTorch](https://github.com/intel/intel-extension-for-pytorch) (IPEX) offers an advanced int8-mixed-bf16 quantization path, which transforms the output of quantized Conv/GEMM operations into the BF16 data type if there is no subsequent quantized operator. This enhancement significantly improves the inference performance of models such as Bert/DistilBert, as the pointwise operators following GEMM will operate with BF16 data instead of FP32.

* Here is the [example code](https://gist.github.com/leslie-fang-intel/acb568f1416c7395b18cd64774aa5a64) for how use this feature with IPEX.
* Please note that this feature may result in accuracy loss for certain models. With IPEX, we have verified its accuracy in models such as Bert, DistilBert, stable diffusion, and some other LLM models. However, we have also observed accuracy issues in models like vision transformers.
* Similarly, we recently recive a feature request in https://github.com/pytorch/pytorch/issues/111487.

### Alternatives

We typically have two options to enable this feature.

### Option 1: Use Autocast
Autocast is naturally employed for BF16 optimization in Inductor. Similarly, we can harness it for PT2E int8-mixed-bf16 features to generate a pattern like `q -> dq -> float32_to_bfloat16 -> conv -> bfloat16_to_fp32 -> q -> dq`. 
* `to_bfloat16` node before conv should be inserted when used Autocast + `torch.compile` together, since conv is in whitelist of Autocast.
* As for inserting `bfloat16_to_fp32` node after conv node, we need to extend the implementation of https://github.com/pytorch/pytorch/blob/93a9b1314b4bc88ccddc0aa438d4d332955027a8/torch/ao/quantization/fx/_decomposed.py#L36-L64 by add these lines at beginning of this function
```
if input.dtype == torch.bfloat16:
  input = input.to(torch.float32)
``` 

Here's an example code snippet:
```
exported_model = capture_pre_autograd_graph(
    model,
    example_inputs
)

# Create X86InductorQuantizer
quantizer = X86InductorQuantizer()
quantizer.set_global(xiq.get_default_x86_inductor_quantization_config())
# PT2E Quantization flow
prepared_model = prepare_pt2e(exported_model, quantizer)
# Calibration
converted_model = convert_pt2e(prepared_model)
torch.ao.quantization.move_exported_model_to_eval(converted_model)

with torch.autocast(device_type="cpu", dtype=torch.bfloat16, enabled=enable_int8_mixed_bf16), torch.no_grad():
	optimized_model = torch.compile(converted_model)

	# Int8-Mixed-BF16 Inference
	quant_output = optimized_model(images)
```

* Pros:
	* Utilize the existing int8-mixed-fp32 quantizer and PT2E flow implementation.
	* Make use of the existing Autocast operator list and mechanism.
* Cons:
	* The Autocast mechanism will convert each input, including the convolution's bias, to BF16. However, for X86InductorQuantizer and the associated Inductor optimization, we anticipate that using float32 for the bias input may yield better accuracy.

### Option 2: Add BFloat16 as a quantization type in PT2E Flow (in QuantizationSpec)
Alternatively, we can introduce BFloat16 as a quantization type in PT2E Flow (within QuantizationSpec). 
* We may need to extend the Observer implementation to annotate its use for int8-mixed-bf16, depending on the quantization recipe.
* During the convert phase, we will examine the observer information to determine if it has been annotated with int8-mixed-bf16. 
	* If the input of a quantization node is in BFloat16 data type, an additional `to_float` node will be inserted before the quantization node.
	* Following the dequantization node, an additional `to_bf16` node will be inserted.

```
exported_model = capture_pre_autograd_graph(
    model,
    example_inputs
)

# Create X86InductorQuantizer
quantizer = X86InductorQuantizer()
quantizer.set_global(xiq.get_default_x86_inductor_quantization_config(dtype=BFloat16))
# PT2E Quantization flow
prepared_model = prepare_pt2e(exported_model, quantizer)
# Calibration
converted_model = convert_pt2e(prepared_model)
torch.ao.quantization.move_exported_model_to_eval(converted_model)

with torch.no_grad():
	optimized_model = torch.compile(converted_model)

	# Int8-Mixed-BF16 Inference
	quant_output = optimized_model(images)
```

* Pros:
	* We can achieve more flexibility with a customized implementation for int8-mixed-bf16 quantization, allowing us to overcome certain limitations in Autocast, such as bias conversion.. 
* Cons:
	* Non-trivial changes may need in QuantizationSpec, Observer, Quantizer and PT2 Flow convert implementation.

We prefer option 1 as it requires fewer changes in the PT2E quantization flow and is clear and straightforward.

### Optimization Inside Inductor
* Conv/GEMM
	Here is the pattern after quantization flow we expect to see in Inductor.
	```
	q -> dq -> float32_to_bfloat16 -> conv -> bfloat16_to_fp32 -> q -> dq
	```
	* Step 1: In the weight prepack phase, `dq -> float32_to_bfloat16 -> conv` will be matched at first to generate a `qconv_bf16_output` node with `int8` input dtype and `bfloat16` output dtype.
	* Step 2: Further more, we will check if `bfloat16_to_fp32 -> q` pattern exists after this `qconv_bf16_output` node. If so, we will further merge `qconv_bf16_output -> bfloat16_to_fp32 -> q` into a `qconv` node with with `int8` input dtype and `int8` output dtype.

* Non-Conv/GEMM.
	* Non-Conv/GEMM pattern will lowering in Inductor CPP Backend for Code Generation.

### Enabling Plans
Here is some plans to follow up the option 1:
* Makes all `onednn.qconv1d_pointwise/linear_pointwise` operators support BF16 output.
* Remove the annotation of output at conv/linear in `X86InductorQuantizer`.
* Extend the decomposed quant to support bf16 input.
* Extend Weight prepack pattern matcher of `dequant -> to_bf16 -> conv/linear`.
* Extend QConv/Linear int8-mixed-bf16 output patterns matcher of `dequant -> to_bf16 -> conv/linear -> to_fp32 -> quant`.
* Extend Postop-passes pattern match of Conv/Linear ReLU/Add/Add_ReLU fusion with FP32/BF16 output.

cc @jerryzh168 @jianyuh @raghuramank100 @jamesr66a @vkuzo @jgong5 @Xia-Weiwen

**Comment by leslie-fang-intel** [comment]
cc @jianan-gu @Xia-Weiwen @jgong5 @jerryzh168

**Comment by jgong5** [comment]
Hi @jerryzh168 This is the proposal I introduced to you during the face to face. Feel free to comment on the options.

**Comment by jerryzh168** [comment]
> Option2: During the convert phase, we will examine the observer information to determine if it has been annotated with int8-mixed-bf16. If the input of a quantization node is in BFloat16 data type, an additional to_float node will be inserted before the quantization node. Following the dequantization node, an additional to_bf16 node will be inserted.

That's not what I have in mind for option 2 actually. What I mean is that "bfloat16" is just another target quantization dtype, while original Tensor are still assumed to be float32. i.e. if user annotates a tensor as following:
```
bfloat16_act_act = QuantizationSpec(dtype=torch.bfloat16, ...)

conv_node.meta[...] = QuantizationAnnotation(input_qspec_map={act: bfloat16_act_qspec, weight:...}, outptu_qspec=bfloat16_act_qspec)
```
quant flow will do:
```
act -> to_bfloat16 (quantize) -> to_fp32 (dequantize) -> conv -> to_bfloat16 (quantize) -> to_fp32 (dequantize)
```

This won't be mixing with int8 dtype conversions.

Or are you saying int8 dtype conversion is actually needed as well?



> Remove the annotation of output at conv/linear in X86InductorQuantizer.

qq: is this only needed for bfloat16 use case or is it needed for all use cases?

**Comment by jerryzh168** [comment]
I feel option 1 is fine if you are OK with having the more coarse grained configurability. there is not much change to the quant flow as you mentioned. although option 2 is also not that complicated either, if what I'm talking about above is what you need for option 2

**Comment by leslie-fang-intel** [comment]
> I feel option 1 is fine if you are OK with having the more coarse grained configurability. there is not much change to the quant flow as you mentioned. although option 2 is also not that complicated either, if what I'm talking about above is what you need for option 2

Thanks for the comment, so I will start to explore more following option 1.

> _qq: is this only needed for bfloat16 use case or is it needed for all use cases?_
  
It's needed for all use cases: fp32 and bfloat16. I will clean up current annotation of output at conv/linear in `X86InductorQuantizer` with following up PRs.

> For option 2: `act -> to_bfloat16 (quantize) -> to_fp32 (dequantize) -> conv -> to_bfloat16 (quantize) -> to_fp32 (dequantize)`

  Some questions about Option 2 you mentioned above: what's the behavior of `to_bfloat16 (quantize)` ? Will it input a Float32 Tensor and output a BFloat16 Tensor.

**Comment by jerryzh168** [comment]
> Some questions about Option 2 you mentioned above: what's the behavior of `to_bfloat16 (quantize)` ? Will it input a Float32 Tensor and output a BFloat16 Tensor.

Yes, to_bfloat16 is the op that quantizes fp32 Tensor to bfloat16 Tensor, it's an "quantize" op, just like existing ones.

**Comment by leslie-fang-intel** [comment]
> > Some questions about Option 2 you mentioned above: what's the behavior of `to_bfloat16 (quantize)` ? Will it input a Float32 Tensor and output a BFloat16 Tensor.
> 
> Yes, to_bfloat16 is the op that quantizes fp32 Tensor to bfloat16 Tensor, it's an "quantize" op, just like existing ones, but with torch.bfloat16 as target dtype.

Ok, but actually in the Inductor optimization, we expect `dequantize -> conv` be fused into a QConv with int8 inputs (as in above **Optimization Inside Inductor** section). If the `quant` outputs a BFloat16 tensor, I think we may not be able to do this optimization.

**Comment by jerryzh168** [comment]
> > > Some questions about Option 2 you mentioned above: what's the behavior of `to_bfloat16 (quantize)` ? Will it input a Float32 Tensor and output a BFloat16 Tensor.
> > 
> > 
> > Yes, to_bfloat16 is the op that quantizes fp32 Tensor to bfloat16 Tensor, it's an "quantize" op, just like existing ones, but with torch.bfloat16 as target dtype.
> 
> Ok, but actually in the Inductor optimization, we expect `dequantize -> conv` be fused into a QConv with int8 inputs (as in above **Optimization Inside Inductor** section). If the `quant` outputs a BFloat16 tensor, I think we may not be able to do this optimization.

I see, if you need int8 inputs then why do you need to insert a bfloat16 op between conv2d and q/dq ops? can't you just do:
`q (int8) -> dq (int8) -> conv2d`, instead of
`q (int8) -> dq (int8) -> to_bfloat16 -> conv2d`?

**Comment by leslie-fang-intel** [comment]
I think because we need to distinguish the pattern of `q (int8) -> dq (int8) -> conv2d` with `q (int8) -> dq (int8) -> to_bfloat16 -> conv2d` for the optimization.

- `dq (int8) -> conv2d` will fused into a QConv with INT8-in and FP32-out.
- `dq (int8) -> to_bfloat16 -> conv2d` will fused into a QConv with INT8-in and BF16-out.

**Comment by jerryzh168** [comment]
> * `dq (int8) -> to_bfloat16 -> conv2d` will fused into a QConv with INT8-in and BF16-out.

but this is not a equivalent transformation, original pattern outputs a fp32 Tensor, how can you transform this to BF16-out?

**Comment by leslie-fang-intel** [comment]
> > * `dq (int8) -> to_bfloat16 -> conv2d` will fused into a QConv with INT8-in and BF16-out.
> 
> but this is not a equivalent transformation, original pattern outputs a fp32 Tensor, how can you transform this to BF16-out?

I think here the original pattern `dq (int8) -> to_bfloat16 -> conv2d` will output BF16 also. The conv2d accepts BF16 input and thus will also output a BF16 tensor in its implementation.

**Comment by jerryzh168** [comment]
I would say in
```
dq (int8) -> to_bfloat16 -> conv2d
```

conv2d doing bfloat16 computation is not in the IR, it's an inferred information about internal implementation details of PyTorch, so it's not a great pattern to use I think

**Comment by jgong5** [comment]
> conv2d doing bfloat16 computation is not in the IR, it's an inferred information about internal implementation details of PyTorch, so it's not a great pattern to use I think

I guess we are already doing similar type inference with the IR under bfloat16 autocast? Do we plan to make everything explicit in the IR with bfloat16 autocast too? Or, we keep it as part of the op semantics - same output dtype as input as the default rule?

**Comment by jerryzh168** [comment]
> Or, we keep it as part of the op semantics - same output dtype as input as the default rule?

how is this used in autocast? you mean autocast also relies on this assumption?

personally I feel it's cleaner to do this more explicitly, and have every aten op to correspond to a single operator instead of relying on these implementation details, since aten ops does not have dtype annotations (unlike edge dialect in executorch). but given the current use cases, this is not too important, and I think it's not too hard to change later, so it's fine to go with relying on the assumption for now as well I think.

**Comment by jgong5** [comment]
> how is this used in autocast? you mean autocast also relies on this assumption? does autocast also work with IR? or is it just working with operators? if it is also working with a graph that contains operators I feel it will also be better to make the semantics of the operators in the graph simpler (just have conv2d op/node to represent one thing, instead of have it represent a context dependent operator/node). if autocast just works with a single operator, then it might be fine I think.

My understanding is that data types are context dependent with autocast and when it is applied to the graph, the type cast would only be inserted to places as needed according to the data types of the inputs and the cast policy, e.g., a few ops like conv/matmul require explicit cast to lower precision and most ops just "fall through" which keep the data types of inputs. When you say "make the semantics of the operators in the graph simpler", do you mean to add explicit type cast ops like "to(bfloat16)" and "to(float)" around all the needed ops just like quantized ops or just mark the data types as metadata on the graph?

**Comment by jerryzh168** [comment]
> > how is this used in autocast? you mean autocast also relies on this assumption? does autocast also work with IR? or is it just working with operators? if it is also working with a graph that contains operators I feel it will also be better to make the semantics of the operators in the graph simpler (just have conv2d op/node to represent one thing, instead of have it represent a context dependent operator/node). if autocast just works with a single operator, then it might be fine I think.
> 
> My understanding is that data types are context dependent with autocast and when it is applied to the graph, the type cast would only be inserted to places as needed according to the data types of the inputs and the cast policy, e.g., a few ops like conv/matmul require explicit cast to lower precision and most ops just "fall through" which keep the data types of inputs. When you say "make the semantics of the operators in the graph simpler", do you mean to add explicit type cast ops like "to(bfloat16)" and "to(float)" around all the needed ops just like quantized ops or just mark the data types as metadata on the graph?

"making the semantics simpler" meaning each aten op, e.g. aten.conv2d should only accept Tensor with the same input dtype, which means we need to add type cast ops "to(bfloat16)", "to(float)" in the graph. I'm not sure if metadata is a part of aten dialect actually, maybe @angelayi @larryliu0820 can help answer this question

**Comment by leslie-fang-intel** [comment]
Close as finished.
